File: Algorithms/C++/Easy/205_Isomorphic_Strings.cpp
class Solution {
public:
    bool isIsomorphic(string s, string t) {
        if (s.empty() && t.empty()) {
            return true;
        }
        if (s.empty() || t.empty()) {
            return false;
        }
        if (s.length() != t.length()) {
            return false;
        }

        unordered_map<char, char> character_map;
        unordered_set<char> character_set;
        for (int i = 0; i < s.length(); i++) {
            if (character_map.find(s[i]) == character_map.end()) {
                if (character_set.find(t[i]) != character_set.end()) {
                    return false;
                }
                character_map[s[i]] = t[i];
                character_set.insert(t[i]);
            } else if (character_map[s[i]] != t[i]) {
                return false;
            }
        }

        return true;
    }
};

File: Algorithms/Swift/Easy/205_Isomorphic_Strings.swift
class Solution {
    func isIsomorphic(s: String, _ t: String) -> Bool {
        if s.isEmpty && t.isEmpty {
            return true
        } else if s.isEmpty || t.isEmpty {
            return false
        } else if s.characters.count != t.characters.count {
            return false
        }

        let tStartIndex = t.characters.startIndex

        var dict = [Character : Character]()
        var set = Set<Character>()

        var mappedS = ""
        for (index, character) in s.characters.enumerate() {
            if let mappedCharacter = dict[character] {
                mappedS += String(mappedCharacter)
            } else {
                let tCharacter = t.characters[tStartIndex.advancedBy(index)]
                if set.contains(tCharacter) {
                    return false
                }

                dict[character] = tCharacter
                set.insert(tCharacter)

                mappedS += String(tCharacter)
            }
        }

        return t == mappedS
    }
}
